src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +74,236 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-                if (File.Exists(fullPathToDotnetFromRoot))
+                AddUnmanagedDllResolver();
+
+                string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
+                if (!string.IsNullOrEmpty(bestSDK))
+                {
+                    yield return bestSDK;
+                }
+
+                string[] dotnetPaths = GetAllAvailableSDKs();
+                // We want to return the newest SDKs first, however, so iterate over the list in reverse order.
+                // If basePath is disqualified because it was later
+                // than the runtime version, this ensures that RegisterDefaults will return the latest valid
+                // SDK instead of the earliest installed.
+                for (int i = dotnetPaths.Length - 1; i >= 0; i--)
                 {
-                    if (!IsWindows)
+                    if (dotnetPaths[i] != bestSDK)
                     {
-                        fullPathToDotnetFromRoot = realpath(fullPathToDotnetFromRoot) ?? fullPathToDotnetFromRoot;
-                        return File.Exists(fullPathToDotnetFromRoot) ? Path.GetDirectoryName(fullPathToDotnetFromRoot) : null;
+                        yield return dotnetPaths[i];
                     }
-
-                    return dotnet_root;
                 }
             }
-
-            return null;
+            finally
+            {
+                RemoveUnmanagedDllResolver();
+            }
         }
 
-        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
-        {
-            string dotnetPath = null;
-            string exeName = IsWindows ? "dotnet.exe" : "dotnet";
+        private static void AddUnmanagedDllResolver() => ModifyUnmanagedDllResolver(loadContext => loadContext.ResolvingUnmanagedDll += HostFxrResolver);
 
-            // First check for the DOTNET_ROOT environment variable, as it's often there as with, for example, dotnet format.
-            if (IntPtr.Size == 4)
+        private static void RemoveUnmanagedDllResolver() => ModifyUnmanagedDllResolver(loadContext => loadContext.ResolvingUnmanagedDll -= HostFxrResolver);
+
+        private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resolverAction)
+        {
+            // For Windows hostfxr is loaded in the process.
+            if (!IsWindows)
             {
-                // 32-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT(x86)", exeName);
+                var loadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+                if (loadContext != null)
+                {
+                    resolverAction(loadContext);
+                }
             }
-            else if (IntPtr.Size == 8)
+        }
+
+        private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
+        {
+            var hostFxrLibName = "libhostfxr";
```

Comment:
I thought you'd implied it could be libhostfxr or hostfxr by having a regex accepting either—was that wrong?

Reply:
I have checked that for osx and linux it always has this name.